src/v/kafka/server/handlers/metadata.cc

```diff
@@ -124,10 +128,23 @@ metadata_response::topic make_topic_response_from_topic_metadata(
         p.partition_index = p_as.id;
         p.leader_id = no_leader;
         auto lt = get_leader_term(tp_ns, p_as.id, md_cache, replicas);
-        if (lt) {
+        if (lt && !is_node_isolated) {
             p.leader_id = lt->leader.value_or(no_leader);
             p.leader_epoch = leader_epoch_from_term(lt->term);
         }
+        if (is_node_isolated) {
+            auto replicas_for_sfuffle = replicas;
+            std::shuffle(
+              replicas_for_sfuffle.begin(),
+              replicas_for_sfuffle.end(),
+              std::default_random_engine());
+            for (const auto& replica : replicas_for_sfuffle) {
+                if (replica != config::node().node_id()) {
+                    p.leader_id = replica;
+                    break;
+                }
```

Comment:
maybe we can replace this shuffle with generating single random number and using it as an index ?

Reply:
I think not
With shuffle we also cover situation with replicas contain only one isolated node. And we should put like no_leader in this case